File: 104.Circle-LineAndCircle-CircleIntersectionAlgo.py
# Circle-Line Intersection: Determines where (if at all) a given straight line crosses a circle.
# Circle-Circle Intersection: Finds the intersection points (if any) between two circles.

# Explanation

# When a line (usually represented as y = mx + c) passes 
# through a circle ((x - a)² + (y - b)² = r²), substituting y gives a quadratic in x.
# If the discriminant (Δ) > 0 → two intersections,
# Δ = 0 → tangent (one intersection),
# Δ < 0 → no intersection.

# Similarly, for two circles, subtract one from another to form a line, 
# then substitute back into one circle equation to get intersection coordinates.

import math

def circle_line_intersection(a, b, r, m, c):
    A = 1 + m**2
    B = 2 * (m * (c - b) - a)
    C = a**2 + (c - b)**2 - r**2

    D = B**2 - 4 * A * C

    if D < 0:
        return []
    
    elif D == 0:
        x = -B / (2 * A)
        y = m * x + c
        return[(x, y)]
    
    else:
        sqrt_D = math.sqrt(D)
        x1 = (-B + sqrt_D) / (2 * A)
        x2 = (-B - sqrt_D) / (2 * A)
        y1 = m * x1 + c
        y2 = m * x2 + c
        return [(x1, y1), (x2, y2)]
    
def circle_circle_intersection(x0, y0, r0, x1, y1, r1):
    dx = x1 - x0
    dy = y1 - y0
    d = math.hypot(dx, dy)

    if d > r0 + r1 or d < abs(r0 - r1):
        return []
    
    a = (r0**2 - r1**2 + d**2) / (2 * d)
    h = math.sqrt(abs(r0**2 - a**2))

    xm = x0 + a * dx / d
    ym = y0 + a * dy / d

    xs1 = xm + h * dy / d
    ys1 = ym - h * dx / d
    xs2 = xm - h * dy / d
    ys2 = ym + h * dx / d

    if h == 0:
        return [(xs1, ys1)]
    
    else:
        return [(xs1, ys1), (xs2, ys2)]
    

# Circle center (0, 0), radius 5
# Line: y = x + 1
line_points = circle_line_intersection(0, 0, 5, 1, 1)
print("Circle-Line Intersection Points:", line_points)

# Circle1: center (0, 0), radius 5
# Circle2: center (4, 0), radius 3
circle_points = circle_circle_intersection(0, 0, 5, 4, 0, 3)
print("Circle-Circle Intersection Points:", circle_points)

File: 104.Circle-LineAndCircle-CircleIntersectionAlgo.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio      */
/* -----------------------------------------------------------------------  */

import java.io.*;
import java.util.*;
import java.lang.Math;

public class Main {

    // ---------------------- Circle-Line Intersection ----------------------
    static List<double[]> circleLineIntersection(double a, double b, double r, double m, double c) {
        // Circle: (x - a)^2 + (y - b)^2 = r^2
        // Line: y = mx + c

        double A = 1 + m * m;                        // Coefficient of x^2
        double B = 2 * (m * (c - b) - a);            // Coefficient of x
        double C = a * a + (c - b) * (c - b) - r * r;// Constant term

        double D = B * B - 4 * A * C;                // Discriminant check

        List<double[]> res = new ArrayList<>();

        if (D < 0) return res;                       // No intersection
        else if (D == 0) {
            double x = -B / (2 * A);
            double y = m * x + c;
            res.add(new double[]{x, y});             // Tangent point
        } else {
            double sqrtD = Math.sqrt(D);
            double x1 = (-B + sqrtD) / (2 * A);
            double x2 = (-B - sqrtD) / (2 * A);
            double y1 = m * x1 + c;
            double y2 = m * x2 + c;
            res.add(new double[]{x1, y1});           // First intersection
            res.add(new double[]{x2, y2});           // Second intersection
        }
        return res;
    }

    // ---------------------- Circle-Circle Intersection ----------------------
    static List<double[]> circleCircleIntersection(double x0, double y0, double r0,
                                                   double x1, double y1, double r1) {
        // Circle 1: (x - x0)^2 + (y - y0)^2 = r0^2
        // Circle 2: (x - x1)^2 + (y - y1)^2 = r1^2

        double dx = x1 - x0;
        double dy = y1 - y0;
        double d = Math.hypot(dx, dy);               // Distance between centers

        List<double[]> res = new ArrayList<>();

        if (d > r0 + r1 || d < Math.abs(r0 - r1)) return res; // No intersection

        double a = (r0 * r0 - r1 * r1 + d * d) / (2 * d);
        double h = Math.sqrt(Math.abs(r0 * r0 - a * a));

        double xm = x0 + a * dx / d;
        double ym = y0 + a * dy / d;

        double xs1 = xm + h * dy / d;
        double ys1 = ym - h * dx / d;
        double xs2 = xm - h * dy / d;
        double ys2 = ym + h * dx / d;

        if (h == 0) res.add(new double[]{xs1, ys1}); // Tangent circles
        else {
            res.add(new double[]{xs1, ys1});         // First intersection
            res.add(new double[]{xs2, ys2});         // Second intersection
        }

        return res;
    }

    // ---------------------- Example Usage ----------------------
    public static void main(String[] args) throws Exception {
        PrintWriter out = new PrintWriter(System.out);

        // Circle center (0, 0), radius 5
        // Line: y = x + 1
        List<double[]> linePoints = circleLineIntersection(0, 0, 5, 1, 1);
        out.println("Circle-Line Intersection Points:");
        for (double[] p : linePoints) out.println(Arrays.toString(p));

        // Circle1: center (0, 0), radius 5
        // Circle2: center (4, 0), radius 3
        List<double[]> circlePoints = circleCircleIntersection(0, 0, 5, 4, 0, 3);
        out.println("Circle-Circle Intersection Points:");
        for (double[] p : circlePoints) out.println(Arrays.toString(p));

        out.flush();
    }
}
